crates/bitcoind_rpc/src/bip158.rs

```diff
@@ -171,27 +154,84 @@ impl<C: RpcApi> Iterator for FilterIter<'_, C> {
     type Item = Result<Event, Error>;
 
     fn next(&mut self) -> Option<Self::Item> {
-        (|| -> Result<_, Error> {
-            // if the next filter matches any of our watched spks, get the block
-            // and return it, inserting relevant block ids along the way
-            self.next_filter()?.map_or(Ok(None), |(block, filter)| {
-                let height = block.height;
-                let hash = block.hash;
-
-                if self.spks.is_empty() {
-                    Err(Error::NoScripts)
-                } else if filter
-                    .match_any(&hash, self.spks.iter().map(|script| script.as_bytes()))
-                    .map_err(Error::Bip158)?
-                {
-                    let block = self.client.get_block(&hash)?;
-                    self.blocks.insert(height, hash);
-                    let inner = EventInner { height, block };
-                    Ok(Some(Event::Block(inner)))
-                } else {
-                    Ok(Some(Event::NoMatch(height)))
+        (|| -> Result<Option<_>, Error> {
+            if self.height > self.stop {
+                return Ok(None);
+            }
+            // Fetch next filter.
+            let mut height = self.height;
+            let mut hash = self.client.get_block_hash(height as _)?;
+            let mut header = self.client.get_block_header(&hash)?;
+
+            // Detect and resolve reorgs: either block at height changed, or its parent changed.
+            let stored_hash = self.blocks.get(&height).copied();
+            let prev_hash = self.blocks.get(&(height - 1)).copied();
```

Comment:
🔧 This is to avoid calling `get` with the underflow height in case `height` is 0.

```suggestion
            let prev_hash = height
                .checked_sub(1)
                .and_then(|height| self.blocks.get(&height).copied());
```